index.html

```diff
@@ -106,6 +108,27 @@ <h2 id="keys-and-identities">Keys and identities</h2>
         </aside>
         <p>The beginning <code>@</code> sign signifies that this represents a public key rather than a message or blob, which start with <code>%</code> and <code>&amp;</code>. Each identity has an associated <a href="#feeds">feed</a>, which is a list of all the messages posted by that identity. This is why the identity is also called a <em>feed ID</em>.</p>
 
+        <h3 id="cryptographic-primitives">Cryptographic primitives</h3>
+
+        <p>The Scuttlebutt protocol relies on NaCl/libsodium's cryptobox primitives. This guide uses the following:</p>
+
+        <dl>
+            <dt><code>nacl_scalarmult(n, p)</code></dt>
+            <dd>This is <a href="https://doc.libsodium.org/advanced/scalar_multiplication">Libsodium's scalar multiplication function</a>, which takes two scalars (usually public and/or secret keys). It has the useful property that, given two key pairs <code>(pk1, sk1)</code> and <code>(pk2, sk2)</code>, <code>nacl_scalarmult(sk1, pk2) == nacl_scalarmult(sk2, pk1)</code>, which allows shared secret derivation between peers who know each other's public key. More on this later.</dd>
+
+            <dt><code>nacl_auth(msg, key)</code> and <code>assert_nacl_auth_verify(authenticator, msg, key)</code></dt>
```

Comment:
We actually just use `nacl_auth()` to create a different message to be signed. It's used like a keyed hash function and I believe `crypto_auth_hmacsha256` would have served the same purpose but this is where we are now.

(Personally I find this easier to read in _my_ Go implementation ([sign](https://github.com/cryptoscope/ssb/blob/master/message/legacy/types.go#L44-L47) & [verify](https://github.com/cryptoscope/ssb/blob/master/message/legacy/verify.go#L71-L74)) but, of course, I'm biased.)



I maybe move this one down to the bottom since it's just used in the HMAC signing mode, which is mostly only useful for testing. (The content stays mostly public but the signatures have a slightly different flavor, so to speak.)

Reply:
> to create a different message to be signed

what message? From what I understand, the result of `nacl_auth` is sent as-is to peers

Reply:
I don't think so. Where would that be? but we might be talking of different things.

The (secret-)handshake uses scalarmult, the mentioned key derivations, box and a hmac using sha512, truncated to the first 32byte.

The only place I know of where `nacl_auth` _is_ used is in the special feed message signing mode.

Reply:
In the "Client Hello" section https://ssbc.github.io/scuttlebutt-protocol-guide/#client-hello

> Client sends (64 bytes)
> 
> ```
> concat(
>   nacl_auth(
>     msg: client_ephemeral_pk,
>     key: network_identifier
>   ),
>   client_ephemeral_pk
> )
```

My understanding is that the result of `concat` is sent directly, with no signature.

Ditto in "Server hello", which are the only two uses of `nacl_auth` I can see in the spec.

Reply:
My bad. `nacl_auth` is implemented as a hmac using sha512, truncated to the first 32byte.

I remembered this wrong and constructed this by hand in my implementation. (Fixed in https://github.com/cryptoscope/secretstream/commit/6c0091d3284e174ed2f67c068f503eb0c893f29a)